### ISSUE TITLE
(typeagead, bug): Model value updated incorrectly when input in { updateOn: "blur" } mode

### ISSUE THREAD

**Issue by AbakumovAlexandr** [ng-bootstrap/ng-bootstrap#2786]
### Bug description:
There are 2 issues when `input`'s `FormControl` is set to `updateOn: "blur"` mode with Typeagead:
```
public formControl = new FormControl(null, { updateOn: "blur" });
```

1. When typing some text (`de`) and then clicking an item (`Delaware`) in the dropdown, it immediately sends `de` to a model. Since the focus is still meant to be on the `input`, this is not what's expected.

2. After that, if you **immediately** give the focus to some other element, the model would be left with `de` which is incorrect since `Delaware` is actually displayed in the `input`. However, if after clicking you give the focus to the `input` itself first and then move it to another element, the model gets correctly updated to `Delaware`.

### Link to minimally-working StackBlitz that reproduces the issue:

[StackBlitz](https://stackblitz.com/edit/ngb-typeahead-blut-not-updatoing).

### Versions of Angular, ng-bootstrap and Bootstrap:

Angular: 6.1.10
ng-bootstrap: 3.3.0
Bootstrap: 4.1.3

**Comment by wannadream** [comment]
I cannot recreate 1 & 2. After I clicked on 'Delaware', value has been synced with model.

> ### Bug description:
> There are 2 issues when `input`'s `FormControl` is set to `updateOn: "blur"` mode with Typeagead:
> 
> ```
> public formControl = new FormControl(null, { updateOn: "blur" });
> ```
> 1. When typing some text (`de`) and then clicking an item (`Delaware`) in the dropdown, it immediately sends `de` to a model. Since the focus is still meant to be on the `input`, this is not what's expected.
> 2. After that, if you **immediately** give the focus to some other element, the model would be left with `de` which is incorrect since `Delaware` is actually displayed in the `input`. However, if after clicking you give the focus to the `input` itself first and then move it to another element, the model gets correctly updated to `Delaware`.
> 
> ### Link to minimally-working StackBlitz that reproduces the issue:
> [StackBlitz](https://stackblitz.com/edit/ngb-typeahead-blut-not-updatoing).
> 
> ### Versions of Angular, ng-bootstrap and Bootstrap:
> Angular: 6.1.10
> ng-bootstrap: 3.3.0
> Bootstrap: 4.1.3

I cannot reproduce 1 & 2 in you StackBlitz. After I clicked on 'Delaware', value has been synced with model.

**Comment by AbakumovAlexandr** [comment]
Just tried at least this part:

> When typing some text (de) and then clicking an item (Delaware) in the dropdown, it immediately sends de to a model.

and it still reproduce-able:

![image](https://user-images.githubusercontent.com/6968009/48010598-83e7ba80-e0e3-11e8-8bf5-e7a0596db3f8.png)

**Comment by wannadream** [comment]
> @wannadream Just tried at least this part:
> 
> > When typing some text (de) and then clicking an item (Delaware) in the dropdown, it immediately sends de to a model.
> 
> and it still reproduce-able:
> 
> ![image](https://user-images.githubusercontent.com/6968009/48010598-83e7ba80-e0e3-11e8-8bf5-e7a0596db3f8.png)

Logically, I did not see anything wrong here. When you select 'Delaware', input value is still 'de'. Hence, form control value is synced to 'de' as blurring event of input triggers. Upon selected item clicked, input value is updated to 'Delaware'. From here, there is no any other 'blur()' event carrying on. Unless, you click on the input and leave it again. To me, it sounds more like a feature, not a bug.

**Comment by AbakumovAlexandr** [comment]
@wannadream My logic is that if user **_clicked_** an option (`Delaware`), this must tell the app that he wants the model to be `Delaware`, not `de`.

Otherwise, he could just enter `de` and not click any option which is working as expected right now.

**Comment by wannadream** [comment]
@AbakumovAlexandr Why don't you change 'updatedOn' to its default behavior 'change'? I am trying to understand the requirement.

**Comment by AbakumovAlexandr** [comment]
@wannadream My requirements are pretty much the same as those which `updateOn: "blur"` mode exists in Angular forms for: I need the model to be updated after an `input` (`typeahead` in this case) has lost the focus (to make it absolutely clear, for the first time, not the second one :) ). And not as a user types every character or selects an option from the list.

This is how any other `input` works in `updateOn: "blur"` mode including `<input type="text">` as well as simple `<select>`. This behavior is natural to Angular forms `input`s in `updateOn: "blur"` mode regardless of `input`'s type.

**Comment by AbakumovAlexandr** [comment]
@pkozlowski-opensource Looks like this, thank you! It fixes #2 from my original post.

Let me just attract your attention to the #1 I still believe needs to be fixed:

> When typing some text (de) and then clicking an item (Delaware) in the dropdown, it immediately sends de to a model. Since the focus is still meant to be on the input, this is not what's expected.

In other words, when in `updateOn: "blur"` mode, `typeahead` should not update the model until the focus is lost.

**Comment by wannadream** [comment]
This is thorny now. `updateOn: "blur"` is from FormControl and it is bound to text input, not part of the `typehead` component. This means a propagation will be required in this scenario. :D

**Comment by dotNetAthlete** [comment]
> @wannadream My logic is that if user **_clicked_** an option (`Delaware`), this must tell the app that he wants the model to be `Delaware`, not `de`.
> 
> Otherwise, he could just enter `de` and not click any option which is working as expected right now.
> 
> The current behavior that the model gets updated to the user-selected value only after he moves the focus from and to the typeahead **twice** just doesn't makes sense.

I would agree with this and extend it further, in that this is probably somehow linked to my list not closing when less than two characters are typed in the input field, despite the logic being there:

`empFormatMatches = (value: any) => value.EmployeeName; 
  searchEmployee = (text$: Observable<string>) =>
    text$.pipe(
      debounceTime(200),
      distinctUntilChanged(),
      tap(() => this.searching = true),
      switchMap(term => term.length < 2 ? [] :
        this.engageService.GetEmployee(term).pipe(
          tap(() => {
            this.searchTerm = term;
            this.searchFailed = false;
          }),
          catchError(() => {
            this.searchFailed = true;
            return of([]);
          }))
      ),
      tap((emp) => {
        this.searching = false;
        if (emp[0].EngageID == "0") {
          this.nonBridgewayCaller = true;
        }
      })
    );`

**Comment by maxokorokov** [comment]
Several `(blur)`-related issues were fixed by #2872 including this one
